File: 1d_wave_nonlin.py
## WAVE EQ
""" this script solves the 1D-wave equations, i.e.
    
        du/dt = -d/dx(1/2*u^2 + g*h - nu*du/dx) + F
        dh/dt = -d/dx(u*h)."""
    
import numpy as np
import matplotlib.pyplot as plt
from scipy import sparse
import time as tictoc

## physical parameters
Lx = 1.     # D = (0,Lx) is the domain
H = 10.      # depth at rest
g = 10.     # gravitational acceleration

## numerical parameters
nx = 500                    # number of grid points for the variable h
nu = nx                   # number of grid points for u
nt = 3000                     # number of time steps to integrate

cfl = 0.9                   # cfl number, 1 is the limit

dx = Lx/nx                  # grid spacing

# staggered grid
x_h = np.arange(dx/2.,Lx,dx)    # grid x on h-points
x_u = np.arange(dx,Lx-dx+dx/2.,dx)    # on u-points

cph = np.sqrt(g*H)          # long wave phase speed
dt = cfl * dx / cph           # dt based on desired cfl-number

# parameters
rho = 1
v = 0.006       # dependent on the grid spacing! choose carefully
F0 = 500.
F = -F0*np.sin(x_h/Lx*8*np.pi)*np.sin(x_h/Lx*np.pi)**2

## OPERATORS
# x gradient on the T-grid
GTx = (sparse.diags(np.ones(nx),0,shape=(nx,nx)) +\
    sparse.diags(-np.ones(nx),-1,shape=(nx,nx))) / dx

GTx[0,-1] = -1/dx   # for periodic boundary conditions

# x gradient on the u-grid
Gux = -GTx.T

# interpolation from u to T and vice versa
ITu = abs(GTx*dx/2.)
IuT = abs(Gux*dx/2.)

## initial conditions
h0 = np.zeros(nx)
u0 = np.zeros(nu)

##  time integration - TRIVIAL WAY
h1 = h0.copy()        # store initial conditions
u1 = u0.copy() 

def rhs(u,h):
    # h is actually eta, h_u is eta+H on the u-grid
    h_u = ITu.dot(h+H)
    
    # non-linear
    rhs_u = -GTx.dot(.5*IuT.dot(u**2) + g*h - v*Gux.dot(u)) + 1/rho/h_u*F0*np.sin(x_h/Lx*8*np.pi - 0.05*it)*np.sin(x_h/Lx*np.pi)**2
    rhs_h = -Gux.dot(u*h_u)
    
    return rhs_u,rhs_h

def RK4(u,h):
    k1 = rhs(u,h)
    k2 = rhs(u + dt/2.*k1[0],h + dt/2.*k1[1])
    k3 = rhs(u + dt/2.*k2[0],h + dt/2.*k2[1])
    k4 = rhs(u + dt*k3[0],h + dt*k3[1])
    
    du = (k1[0] + 2*k2[0] + 2*k3[0] + k4[0]) / 6.
    dh = (k1[1] + 2*k2[1] + 2*k3[1] + k4[1]) / 6.
    return du,dh

fig,(ax1,ax2,ax3) = plt.subplots(3,1,figsize=(9,8))

mass1 = np.empty(nt)
energy1 = np.empty(nt)

time = np.arange(nt)*dt

for it in range(nt):
    
    mass1[it] = (h1+H).sum()/(H*nx)
    energy1[it] = (.5*(h1+H)*IuT.dot(u1**2) + .5*h1**2).mean()
    
    if it == 0:
        l11, = ax1.plot(x_h,h1)
        ax1.set_ylim(H-2,H+2)
        ax1.set_xlabel('x')
        ax1.set_ylabel(r'$h$')
        ax1.set_title(r'surface displacement $h$')
        
        l21, = ax2.plot(time[:it],mass1[:it])
        ax2.set_ylim(0.95,1.05)
        ax2.plot([0,nt*dt],[mass1[0],mass1[0]],'grey',lw=0.5)
        ax2.set_xlabel('time')
        ax2.set_ylabel('mass')
        
        l31, = ax3.plot(time[:it],energy1[:it])
        ax3.set_ylim(0,4)
        ax3.set_xlim(ax2.get_xlim())
        ax3.set_xlabel('time')
        ax3.set_ylabel('energy')
        
        plt.tight_layout()
        
    else:
        du1,dh1 = RK4(u1,h1)
        u1 += dt*du1
        h1 += dt*dh1
        
        if it % 5 == 1:
            plt.pause(0.00001)
            l11.set_data(x_h,h1+H)
            l21.set_data(time[:it],mass1[:it])
            l31.set_data(time[:it],energy1[:it])
            
plt.pause(2)
plt.close(fig)
        
        

File: wave/1d_wave_nonlin.jl
# define constants
using PyPlot

const g = 10.
const H = 10.
const N = 500
const Nt = 3000
const cfl = 0.9
const rho = 1.
const ν = 0.006
const F0 = 500.
const Lx = 1.
const dx = Lx/N
const dxinv = 1./dx
const cph = sqrt(g*H)
const dt = cfl * dx / cph
const RKα = [1/6.,1/3.,1/3.,1/6.]
const RKβ = [0.5,0.5,1.]

# grid
const x_h = dx/2:dx:Lx
const x_u = dx:dx:Lx        # no u-point at x=0 but at x=L (periodicity)

F(x,t) = -F0*sin.(8π*x/Lx + 200*t).*sin.(π*x/Lx).^2/rho

# OPERATORS
function Gux(res,u)
    res[1] = dxinv*(u[1]-u[end])
    res[2:end] = dxinv*(u[2:end]-u[1:end-1])
    return res
end

function GTx(res,η)
    res[1:end-1] = dxinv*(η[2:end]-η[1:end-1])
    res[end] = dxinv*(η[1]-η[end])
    return res
end

function ITu(res,η)
    res[1:end-1] = .5*(η[1:end-1] + η[2:end])
    res[end] = .5*(η[1]+η[end])
    return res
end

function IuT(res,u)
    res[1] = .5*(u[1] + u[end])
    res[2:end] = .5*(u[1:end-1]+u[2:end])
    return res
end


# initial conditions
η0 = zeros(N)
u0 = zeros(N)

function rhs(du,dη,h_u,u_h,dudx,u,η,t)
    h_u = ITu(h_u,η+H)
    u_h = IuT(u_h,u.^2)
    dudx = Gux(dudx,u)

    du = -GTx(du,.5*u_h + g*η - ν*dudx)
    du += F(x_h,t)./h_u

    dη = -Gux(dη,u.*h_u)

    return du,dη
end

function time_integration(Nt,u,η)

    # pre-allocate memory
    u0,η0 = zeros(u),zeros(η)
    u1,η1 = zeros(u),zeros(η)
    du,dη = zeros(u),zeros(η)
    h_u = zeros(u)
    u_h = zeros(η)
    dudx = zeros(η)
    t = 0.

    # for output
    u_out = zeros(Nt+1,length(u))
    η_out = zeros(Nt+1,length(η))

    # store the initial conditions
    u_out[1,:] = u
    η_out[1,:] = η

    for i = 1:Nt
        u1[:] = u
        η1[:] = η

        for RKi = 1:4
            du,dη = rhs(du,dη,h_u,u_h,dudx,u1,η1,t)

            if RKi < 4 # RHS update for the next RK-step
                u1 = u + RKβ[RKi]*dt*du
                η1 = η + RKβ[RKi]*dt*dη
            end

            # Summing all the RHS on the go
            u0 += RKα[RKi]*dt*du
            η0 += RKα[RKi]*dt*dη

        end

        u[:] = u0
        η[:] = η0
        t += dt

        # # output
        u_out[i+1,:] = u
        η_out[i+1,:] = η

    end
    u_out,η_out
end

time_integration(1,u0,η0)
@time u,η = time_integration(Nt,u0,η0)

## PLOTTING
fig,ax = subplots()

l1, = ax[:plot](x_h,η[1,:])
ax[:set_ylim](-2,2)
ax[:set_xlabel]("x")
ax[:set_ylabel]("η")
tight_layout()

for it = 2:5:Nt+1
    pause(0.00001)
    l1[:set_data](x_h,η[it,:])
end

pause(2)
close(fig)
